### PR TITLE
Server issues

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleHierarchyDownloader.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      treeResponse.getComponentsList().forEach(c -> modulesWithPath.put(c.getKey(), findPathFromRoot(c, ancestors)));
+
+      return modulesWithPath;
+
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to load module hierarchy", e);
+    }
+  }
+
+  private String findPathFromRoot(Component component, Map<Component, Component> ancestors) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleHierarchyDownloader.java]: MAJOR Make "findPathFromRoot" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleHierarchyDownloader.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    return path.toString();
+  }
+
+  @CheckForNull
+  private String fetchAncestorId(String moduleId) throws IOException {
+    WsResponse response = wsClient.get("api/components/show.protobuf?id=" + StringUtils.urlEncode(moduleId));
+    try (InputStream stream = response.contentStream()) {
+      ShowWsResponse showResponse = WsComponents.ShowWsResponse.parseFrom(stream);
+      return showResponse.getAncestorsList().stream().map(c -> c.getId()).findFirst().orElse(null);
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleHierarchyDownloader.java]: MINOR Replace this lambda with a method reference. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

File: core/src/main/java/org/sonarsource/sonarlint/core/util/ParserIterator.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        input.close();
+      } catch (IOException e) {
+        throw new IllegalStateException("Error closing stream", e);
+      }
+    }
+    return el;
+  }
+
+  @Override
+  public T next() {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/util/ParserIterator.java]: MINOR Add a "NoSuchElementException" for iteration beyond the end of the collection. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2272)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/storage/StorageGlobalContainer.java [hunk 4]
```diff
@@ -22,6 +22,8 @@
 import java.text.SimpleDateFormat;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/storage/StorageGlobalContainer.java]: MINOR Remove this unused import 'java.util.List'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MAJOR 2 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/objectstore/ObjectStore.java [hunk 5]
```diff
@@ -30,7 +31,7 @@
  */
 public interface ObjectStore<K, V> {
 
-  void write(K key, V value) throws IOException;
+  void write(K key, Iterator<V> values) throws IOException;
 
-  Optional<V> read(K key) throws IOException;
+  Optional<Iterator<V>> read(K key) throws IOException;
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/objectstore/ObjectStore.java]: Can't we introduce the iterator at another level of abstraction? Either in `SimpleObjectStore` or in `ServerIssueStore`?

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java [hunk 6]
```diff
@@ -120,6 +117,11 @@ private void updateStatus(Path temp) {
     ProtobufUtil.writeToFile(updateStatus, temp.resolve(StorageManager.UPDATE_STATUS_PB));
   }
 
+  private void fetchModuleHierarchy(String moduleKey, ModuleConfiguration.Builder builder) {
+    Map<String, String> moduleHierarchy = moduleHierarchyDownloader.fetchModuleHierarchy(moduleKey);
+    moduleHierarchy.forEach((key, path) -> builder.getMutableModulesKeysByPath().put(path, key));
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java]: Wouldn't it be better to extract `builder.getMutableModulesKeysByPath()` to a variable before the `forEach` ?

File: core/src/main/java/org/sonarsource/sonarlint/core/container/storage/IssueStoreGetter.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    Map<String, String> moduleKeysByPath = moduleConfig.getModulesKeysByPath();
+
+    // find longest prefix match
+    String subModuleKey = moduleKey;
+    String relativeFilePath = filePath;
+
+    for (Map.Entry<String, String> e : moduleKeysByPath.entrySet()) {
+      if (filePath.startsWith(e.getKey()) && (relativeFilePath == null || relativeFilePath.length() < e.getKey().length())) {
+        subModuleKey = e.getValue();
+        relativeFilePath = e.getKey();
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/storage/IssueStoreGetter.java]: How about extracting `e.getKey()` to a variable?

File: core/src/main/java/org/sonarsource/sonarlint/core/container/storage/IssueStoreGetter.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  private String getFileKey(String moduleKey, String filePath) {
+    ModuleConfiguration moduleConfig = storageManager.readModuleConfigFromStorage(moduleKey);
+    Map<String, String> moduleKeysByPath = moduleConfig.getModulesKeysByPath();
+
+    // find longest prefix match
+    String subModuleKey = moduleKey;
+    String relativeFilePath = filePath;
+
+    for (Map.Entry<String, String> e : moduleKeysByPath.entrySet()) {
+      if (filePath.startsWith(e.getKey()) && (relativeFilePath == null || relativeFilePath.length() < e.getKey().length())) {
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/storage/IssueStoreGetter.java]: When would `relativeFilePath` be `null`?

File: core/src/test/java/org/sonarsource/sonarlint/core/container/storage/IssueStoreGetterTest.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+import com.google.common.base.Objects;
+
+public class IssueStoreGetterTest {
+  private IssueStoreGetter issueStoreGetter;
+  private IssueStore issueStore;
+  private StorageManager storage;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
```
Comment (SonarTech) [inline core/src/test/java/org/sonarsource/sonarlint/core/container/storage/IssueStoreGetterTest.java]: MINOR Remove this unused "TemporaryFolder". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2924)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java [hunk 10]
```diff
@@ -120,6 +117,11 @@ private void updateStatus(Path temp) {
     ProtobufUtil.writeToFile(updateStatus, temp.resolve(StorageManager.UPDATE_STATUS_PB));
   }
 
+  private void fetchModuleHierarchy(String moduleKey, ModuleConfiguration.Builder builder) {
+    Map<String, String> moduleHierarchy = moduleHierarchyDownloader.fetchModuleHierarchy(moduleKey);
+    moduleHierarchy.forEach((key, path) -> builder.getMutableModulePathByKey().put(key, path));
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleConfigUpdateExecutor.java]: MINOR Replace this lambda with a method reference. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

**Comment by SonarTech** [comment]
SonarQube analysis reported 9 issues
* MAJOR 1 major
* MINOR 7 minor
* INFO 1 info

Watch the comments in this conversation to review them.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/objectstore/ObjectStore.java (hunk 5 above)
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/objectstore/ObjectStore.java]: well seen!

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MINOR 1 minor
* INFO 1 info

Watch the comments in this conversation to review them.

**Pull request merged**